pageserver/src/tenant/secondary/heatmap_uploader.rs

```diff
@@ -0,0 +1,529 @@
+                }
+            }
+        }
+
+        // Stale attachments do not upload anything: if we are in this state, there is probably some
+        // other attachment in mode Single or Multi running on another pageserver, and we don't
+        // want to thrash and overwrite their heatmap uploads.
+        if tenant.get_attach_mode() == AttachmentMode::Stale {
+            return;
+        }
+
+        // Create an entry in self.tenants if one doesn't already exist: this will later be updated
+        // with the completion time in on_completion.
+        let state = self
+            .tenants
+            .entry(*tenant.get_tenant_shard_id())
+            .or_insert_with(|| UploaderTenantState {
+                tenant: Arc::downgrade(&tenant),
+                last_upload: None,
+                next_upload: Some(Instant::now()),
+                last_digest: None,
+            });
+
+        // Decline to do the upload if insufficient time has passed
+        if state.next_upload.map(|nu| &nu > now).unwrap_or(false) {
+            return;
+        }
+
+        let last_digest = state.last_digest;
+        self.tenants_pending.push_back(UploadPending {
+            tenant,
+            last_digest,
+        })
+    }
+
+    fn spawn_pending(&mut self) {
+        while !self.tenants_pending.is_empty()
+            && self.tenants_uploading.len() < self.concurrent_uploads
+        {
+            // unwrap: loop condition includes !is_empty()
+            let pending = self.tenants_pending.pop_front().unwrap();
+            self.spawn_upload(pending.tenant, pending.last_digest);
+        }
+    }
+
+    fn spawn_upload(&mut self, tenant: Arc<Tenant>, last_digest: Option<md5::Digest>) {
+        let remote_storage = self.remote_storage.clone();
+        let tenant_shard_id = *tenant.get_tenant_shard_id();
+        let (completion, barrier) = utils::completion::channel();
+        let result_tx = self.task_result_tx.clone();
+        self.tasks.spawn(async move {
+            // Guard for the barrier in [`WriteInProgress`]
+            let _completion = completion;
+
+            let digest = match upload_tenant_heatmap(remote_storage, &tenant, last_digest)
+                .instrument(tracing::info_span!(
+                    "upload_heatmap",
+                    tenant_id = %tenant.get_tenant_shard_id().tenant_id,
+                    shard_id = %tenant.get_tenant_shard_id().shard_slug()
+                ))
+                .await
+            {
+                Ok(digest) => digest,
+                Err(e) => {
+                    tracing::warn!(
+                        "Failed to upload heatmap for tenant {}: {e:#}",
+                        tenant.get_tenant_shard_id(),
+                    );
+                    last_digest
+                }
+            };
+
+            let next_upload = tenant
+                .get_heatmap_period()
+                .and_then(|period| Instant::now().checked_add(period));
+
```

Comment:
1. Redundant now() call fixed in 2f8632572
2. Because heatmap uploads are small S3 PUTs vs. an expected multi-minute interval, I think we don't need to do an overrun check similar to compaction/gc.  Checking on the scheduling latency would be more interesting, but it's not trivial to do in a non-spammy way, because it's quite likely that right after startup we'll take a while to get everything uploaded, as heatmap uploads are intentionally run at a limited concurrency.

I would like some visibility of how long it takes though: I've added a duration histogram in d43255e41 along with an error counter.

Reply:
> I think we don't need to do an overrun check similar to compaction/gc

The benefit of using that `warn_when_period_overrun` function is that it has a standardized metric, and I have a [dashboard visualizing it.](https://neonprod.grafana.net/d/f77f7512-57d2-4223-a0c2-edde69cb6f73/background-loop-period-overrun?orgId=1)

Reply:
Hmm, fair point.  Let me pick this up in the next PR  (#6123 ), where I'm generalizing the scheduler: I will hook it in to that existing metric, or something equally generic for tasks scheduled in this way.